clang/test/CodeGenSYCL/device_has.cpp

```diff
@@ -8,25 +8,30 @@ queue q;
 
 // CHECK: define dso_local spir_kernel void @{{.*}}kernel_name_1{{.*}} !sycl_declared_aspects ![[ASPECTS1:[0-9]+]] !srcloc ![[SRCLOC1:[0-9]+]]
 
-// CHECK: define dso_local spir_func void @{{.*}}func1{{.*}} !sycl_declared_aspects ![[ASPECTS1]] !srcloc ![[SRCLOC2:[0-9]+]] {
+// CHECK: define dso_local spir_func void @{{.*}}func1{{.*}} !sycl_declared_aspects ![[ASPECTS1]] !srcloc ![[SRCLOC2:[0-9]+]]
+// CHECK-SAME: !sycl_used_aspects
```

Comment:
IIUC the aspect list should be the same as that in !sycl_decl_aspects right? Please add ![[ASPECT1]] to this CHECK as well, and make similar modifications to remaining CHECK lines.

Reply:
Added e647f36.
About your question. Since `sycl_declared_aspects` strictly requires the device to support declared aspects whether they are really used or no. I believe It's ok to merge them into `sycl_used_aspects` so we treat them as used. @steffenlarsen am I right?

Reply:
My point was that you lose information about which aspects were actually used in the kernel once this new pass runs. I am not sure for what all purposes actual aspect usage information is required, but off the top of my head I expect it is required to emit proper diagnostics, etc. I am assuming that this merge is done only after any pass which requires knowledge of actual aspects used is complete?

Reply:
Yes, we perform this merge at the end of the pass.

Reply:
Since the merge happens after the verification it should be okay, though you could argue that if functions disagree about aspects in their `device_has` attributes (or properties) it should also be a warning. That said, the spec only says a diagnostic needs to be issued if a kernel with a `device_has` "uses" an optional kernel feature.

After the pass, I don't see any problem in considering all `device_has` aspects as propagated uses as in some ways that is how the runtime will see them too. However, maybe there could be value in preserving information about the origin of the aspects, i.e. types/functions explicitly marked as using an aspect such as `sycl::half`. We could do this by changing the name of the metadata produced from the `__sycl_detail__::__uses_aspects__` attribute, then in this pass we would extract that information and pass it along as we currently do the `sycl_used_aspects`. However, I don't know of a current use-case for this so I don't think it has to be part of these changes.